### PR TITLE
feat(processors.dedup): Add state persistence between runs

### LINKED ISSUE #14059: Dedup plugin save state
### Use Case

Keep the dedup plugin deduping across config reloads.

### Expected behavior

Dedup plugin only emits metrics when the metric changes, even after the config gets reloaded.

### Actual behavior

Config reloading causes metrics to be emitted again, even if there were no changes.

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by gazpachoking**
- [x] Updated associated README.md.
- [ ] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #14059

This adds the state persistence interface to the processors.dedup plugin. It's implemented by using the influx serializer/parser to store the metrics.

Turns out there was a bug that non-streaming processors couldn't have state persistence, also fixed that issue in here to make it work.

### CODE REVIEW

**Review by srebhan — APPROVED** [review]
Very nice addition @gazpachoking! Thanks for sharing!

File: agent/agent.go [hunk 1]
```diff
@@ -4,6 +4,7 @@ import (
 	"context"
 	"errors"
 	"fmt"
+	"github.com/influxdata/telegraf/plugins/processors"
```
Comment (Hipska) [inline agent/agent.go:7]: Move this line to the block below with the other telegraf imports

File: plugins/processors/dedup/dedup.go [hunk 2]
```diff
@@ -3,6 +3,9 @@ package dedup
 
 import (
 	_ "embed"
+	"fmt"
+	"github.com/influxdata/telegraf/plugins/parsers/influx"
+	influxSerializer "github.com/influxdata/telegraf/plugins/serializers/influx"
```
Comment (Hipska) [inline plugins/processors/dedup/dedup.go:8]: Same here, keep telegraf imports together.

Maybe also rename the first one to `influxParser` to have no confusion at all in the code?

**Review by Hipska — CHANGES_REQUESTED** [review]
@srebhan should a `StatefulPlugin` have tests on their `GetState` and `SetState` methods?

File: plugins/processors/dedup/dedup.go [hunk 3]
```diff
@@ -117,6 +120,32 @@ func (d *Dedup) Apply(metrics ...telegraf.Metric) []telegraf.Metric {
 	return metrics
 }
 
+func (d *Dedup) GetState() interface{} {
+	s := &influxSerializer.Serializer{}
+	v := make([]telegraf.Metric, 0, len(d.Cache))
+	for _, value := range d.Cache {
+		v = append(v, value)
+	}
```
Comment (Hipska) [inline plugins/processors/dedup/dedup.go:128]: Is this copy needed?

**Comment by Hipska** [comment]
@gazpachoking Is there really a need to serialise and unserialise the metrics? Can't you just return `d.Cache`?

**Comment by srebhan** [comment]
@Hipska it's really better to serialize the metrics as the persister will JSON-serialize the returned state... :-/

**Comment by Hipska** [comment]
Indeed, and now it will make a JSON string of the serialised metric. So first parse the JSON and then parse the metric. I don't see the benefit here?

**Comment by srebhan** [comment]
@Hipska the benefit is that the metric is serialized. :-P If you pass a metric to the JSON serializer it will do nothing as the fields are not exported...

**Comment by Hipska** [comment]
Oh, then maybe that (make metrics JSON serialisable) should be done then?

**Comment by srebhan** [comment]
@Hipska why? You are losing information in serializing to JSON e.g. if a number is an integer or a float. What is better if we serialize to JSON than if we serialize to line-protocol?

**Comment by gazpachoking** [comment]
> the benefit is that the metric is serialized. :-P If you pass a metric to the JSON serializer it will do nothing as the fields are not exported...

Yeah, I tried just returning `d.Cache` first, and was just getting an empty JSON object out. I couldn't figure out how to make Go JSON serialize the metric, (and didn't even think of the int vs float issue,) so I went with one of the existing ways to serialize metrics.

**Pull request merged**